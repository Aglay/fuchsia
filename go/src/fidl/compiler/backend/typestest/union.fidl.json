{
  "version": "0.0.1",
  "name": "test.name",
  "library_dependencies": [],
  "const_declarations": [],
  "enum_declarations": [],
  "interface_declarations": [],
  "struct_declarations": [],
  "table_declarations": [],
  "union_declarations": [
    {
      "name": "test.name/Union",
      "members": [
        {
          "type": {
            "kind": "primitive",
            "subtype": "int32"
          },
          "name": "Primitive",
          "size": 4,
          "max_out_of_line": 0,
          "alignment": 4,
          "offset": 8
        },
        {
          "type": {
            "kind": "string",
            "nullable": false
          },
          "name": "StringNeedsConstructor",
          "size": 16,
          "max_out_of_line": 4294967295,
          "alignment": 8,
          "offset": 8
        },
        {
          "type": {
            "kind": "vector",
            "element_type": {
              "kind": "string",
              "nullable": false
            },
            "nullable": false
          },
          "name": "VectorStringAlsoNeedsConstructor",
          "size": 16,
          "max_out_of_line": 4294967295,
          "alignment": 8,
          "offset": 8
        }
      ],
      "size": 24,
      "max_out_of_line": 4294967295,
      "alignment": 8,
      "max_handles": 0
    }
  ],
  "declaration_order": [
    "test.name/Union"
  ],
  "declarations": {
    "test.name/Union": "union"
  }
}
