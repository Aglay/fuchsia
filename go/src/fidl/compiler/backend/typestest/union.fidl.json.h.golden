// WARNING: This file is machine generated by fidlgen.

#pragma once

#include "lib/fidl/cpp/internal/header.h"


namespace test {
namespace name {

class Union;

class Union {
 public:
  Union();
  ~Union();

  Union(Union&&);
  Union& operator=(Union&&);

  enum class Tag : fidl_union_tag_t {
    kPrimitive = 0,
    kStringNeedsConstructor = 1,
    kVectorStringAlsoNeedsConstructor = 2,
    Invalid = ::std::numeric_limits<::fidl_union_tag_t>::max(),
  };

  static inline ::std::unique_ptr<Union> New() { return ::std::make_unique<Union>(); }

  void Encode(::fidl::Encoder* encoder, size_t offset);
  static void Decode(::fidl::Decoder* decoder, Union* value, size_t offset);
  zx_status_t Clone(Union* result) const;

  bool has_invalid_tag() const { return tag_ == ::std::numeric_limits<::fidl_union_tag_t>::max(); }

  bool is_Primitive() const { return tag_ == 0; }
  
  int32_t& Primitive() {
    EnsureStorageInitialized(0);
    return Primitive_;
  }
  
  const int32_t& Primitive() const { return Primitive_; }
  void set_Primitive(int32_t value);

  bool is_StringNeedsConstructor() const { return tag_ == 1; }
  
  ::fidl::StringPtr& StringNeedsConstructor() {
    EnsureStorageInitialized(1);
    return StringNeedsConstructor_;
  }
  
  const ::fidl::StringPtr& StringNeedsConstructor() const { return StringNeedsConstructor_; }
  void set_StringNeedsConstructor(::fidl::StringPtr value);

  bool is_VectorStringAlsoNeedsConstructor() const { return tag_ == 2; }
  
  ::fidl::VectorPtr<::fidl::StringPtr>& VectorStringAlsoNeedsConstructor() {
    EnsureStorageInitialized(2);
    return VectorStringAlsoNeedsConstructor_;
  }
  
  const ::fidl::VectorPtr<::fidl::StringPtr>& VectorStringAlsoNeedsConstructor() const { return VectorStringAlsoNeedsConstructor_; }
  void set_VectorStringAlsoNeedsConstructor(::fidl::VectorPtr<::fidl::StringPtr> value);

  Tag Which() const { return Tag(tag_); }

 private:
  friend bool operator==(const Union& lhs, const Union& rhs);
  void Destroy();
  void EnsureStorageInitialized(::fidl_union_tag_t tag);

  ::fidl_union_tag_t tag_;
  union {
    int32_t Primitive_;
    ::fidl::StringPtr StringNeedsConstructor_;
    ::fidl::VectorPtr<::fidl::StringPtr> VectorStringAlsoNeedsConstructor_;
  };
};

bool operator==(const Union& lhs, const Union& rhs);
inline bool operator!=(const Union& lhs, const Union& rhs) {
  return !(lhs == rhs);
}

inline zx_status_t Clone(const ::test::name::Union& value,
                         ::test::name::Union* result) {
  return value.Clone(result);
}

using UnionPtr = ::std::unique_ptr<Union>;
}  // namespace name
}  // namespace test
namespace fidl {

template <>
struct CodingTraits<::test::name::Union>
    : public EncodableCodingTraits<::test::name::Union, 24> {};

inline zx_status_t Clone(const ::test::name::Union& value,
                         ::test::name::Union* result) {
  return ::test::name::Clone(value, result);
}}  // namespace fidl
